File: scripts/discover.py
#!/usr/bin/env python3
"""Propose what belongs on the page, instead of remembering it.

The first seed list was hand-written from memory, which is how a list is wrong the week after it is
made. Membership is now derived and only vetoed by hand:

    candidate pool  ->  a bar anyone can check  ->  a layer  ->  human review

The pool comes from lists other people maintain full time and from GitHub's own search. The bar is
deliberately dull — alive, adopted, open source, not a tutorial — because every interesting judgement
belongs in the measured columns, not in a private opinion about what deserves to be listed.

Usage: GITHUB_TOKEN=... python3 scripts/discover.py [--stars 1000]
"""
from __future__ import annotations

import json
import os
import re
import sys
import time
from concurrent.futures import ThreadPoolExecutor
import urllib.error
import urllib.parse
import urllib.request
from datetime import UTC, datetime
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
NOW = datetime.now(UTC)
API = "https://api.github.com"

# Curated by people who do it full time. Sampled 2026-09-13: 17/25 and 13/25 of their links had been
# pushed within 30 days, against 5/24 for e2b-dev/awesome-ai-agents, which is excluded for that reason.
SOURCE_LISTS = ["hesreallyhim/awesome-claude-code", "kyrolabs/awesome-agents"]

# One query set per layer. The layer a project lands in is decided by the query that found it and by
# the keywords below, so a project can be proposed for a layer without anyone having heard of it.
# Agents are split by what the agent is for, not by "coding versus general". The earlier split made
# "general" a bucket for everything that was not coding: research agents, browser agents, desktop
# assistants and vertical apps all landed together, and products like openclaw sit on both sides of it.
LAYERS = {
    "coding-agents": (
        "Coding agents",
        "You use it to write and change code. You are its user, not a builder on it.",
        ["topic:coding-agent", '"coding agent" in:name,description', "topic:ai-coding-assistant"],
        r"coding agent|code agent|pair program|cli agent|ide|editor|swe|vibe coding|codebase",
    ),
    "research-agents": (
        "Research and knowledge agents",
        "Agents that go and find out: deep research, document analysis, report writing.",
        ["deep research agent in:name,description", "research agent in:name,description", "document analysis agent in:description"],
        r"research|deep.?research|report|literature|analys[ei]s|knowledge base|summari",
    ),
    "browser-agents": (
        "Browser and computer-use agents",
        "Agents that drive a screen: browsers, desktops, phones.",
        ["browser agent in:name,description", "computer use agent in:description", "gui agent in:name,description"],
        r"browser|computer.use|gui agent|web automation|click|screenshot|desktop automation",
    ),
    "assistant-agents": (
        "Personal and desktop assistants",
        "The agent that sits next to your work and does errands across your own tools.",
        ["personal ai assistant in:name,description", "self-hosted assistant in:description", "desktop ai agent in:description"],
        r"personal|assistant|self.hosted|desktop|inbox|calendar|workspace|companion",
    ),
    "clients": (
        "Clients and interfaces",
        "The chat window, terminal UI or desktop shell you put in front of an agent.",
        ["ai chat ui in:name,description", "llm client in:name,description", "agent ui in:name,description"],
        r"\bui\b|client|chat(bot)? (app|interface|ui)|frontend|desktop app|tui|web ui",
    ),
    "harness": (
        "Harnesses and agent runtimes",
        "The loop you build on: tool execution, context management, turn control. Opinionated, not a library.",
        ["agent harness in:name,description", "agent runtime in:name,description", "agent loop in:description"],
        r"harness|agent loop|runtime|toolkit|agent os|execution engine",
    ),
    "frameworks": (
        "Frameworks and SDKs",
        "Composable libraries for building an agent: handoffs, structured output, graphs.",
        ["topic:agent-framework", "topic:ai-agents", "agent sdk in:name,description"],
        r"framework|sdk|library|multi.?agent|graph|orchestrat|workflow.*(build|defin)",
    ),
    "skills": (
        "Skills and instruction assets",
        "Files you load into an agent: skills, prompt packs, AGENTS.md collections. No code to run.",
        ["topic:claude-code", "agent skills in:name,description", "topic:prompt-engineering stars:>3000"],
        r"skill|prompt|instruction|agents\.md|claude\.md|playbook|persona|rules",
    ),
    "orchestration": (
        "Orchestration, durability and deployment",
        "What keeps an agent alive between turns: queues, retries, state, serving.",
        ["durable execution in:description", "workflow engine in:description stars:>2000", "agent platform in:description"],
        r"durable|queue|retry|state machine|serving|deploy|platform|scheduler",
    ),
    "memory": (
        "Memory and context",
        "What the agent remembers after the session ends.",
        ["agent memory in:name,description", "topic:memory stars:>2000", "context engineering in:description"],
        r"memory|context|recall|knowledge graph|retriev|rag\b",
    ),
    "gateways": (
        "Model gateways, routing and account switching",
        "Everything between your agent and whichever model answers: proxies, routers, key and account switchers.",
        ["llm gateway in:name,description", "llm proxy in:name,description", "model router in:description", "api key switcher in:description"],
        r"gateway|proxy|router|switch(er)?|load balanc|fallback|multi.?provider|rate limit|token.*(cost|budget)|api key",
    ),
    "protocol": (
        "MCP, protocols and registries",
        "The standards agents speak, the SDKs that implement them, and the registries tools are published to.",
        ["topic:mcp stars:>2000", "topic:model-context-protocol", "agent2agent OR a2a protocol in:name,description", "mcp registry in:name,description"],
        r"\bmcp\b|model context protocol|protocol|registry|a2a|interop|spec\b|standard",
    ),
    "capability-tools": (
        "Capability tools",
        "What an agent reaches for: search, scraping, documents, data access.",
        ["web scraping ai in:description", "agent tools in:name,description", "llm search api in:description"],
        r"scrap|crawl|search|document|spreadsheet|extract|data access|api wrapper",
    ),
    "sandboxes": (
        "Sandboxes and execution",
        "Where the agent's code is allowed to run.",
        ["code interpreter sandbox in:description", "topic:sandbox stars:>2000", "agent sandbox in:description"],
        r"sandbox|isolat|container|microvm|execution environment|code interpreter",
    ),
    "evaluation": (
        "Evaluation and observability",
        "How you find out whether the agent is still doing its job.",
        ["topic:llm-evaluation", "llm observability in:name,description", "agent evaluation in:name,description"],
        r"eval|observab|tracing|monitor|benchmark|telemetry|experiment",
    ),
    "security": (
        "Security and guardrails",
        "Stopping an agent from being talked into something, or taking the system with it.",
        ["prompt injection in:name,description", "llm guardrails in:name,description", "agent security in:description"],
        r"injection|guardrail|security|permission|red team|scanner|governance|policy",
    ),
}

# The layer keywords are loose on purpose, so relevance needs its own anchor: without it the first run
# proposed The Powder Toy, a Java interview guide and MockServer, all of which matched "sandbox",
# "agent" and "proxy" honestly enough.
DOMAIN = re.compile(r"\b(agent|agentic|llm|ai|mcp|prompt|model context|claude|gpt|codex|copilot|openai|anthropic|rag)\b", re.I)

# A list, a course or a demo is not a project you put in production.
EXCLUDE = re.compile(r"awesome|tutorial|course|roadmap|handbook|cookbook|examples?$|demo|starter|template|boilerplate|papers?$|interview|study|learn", re.I)


def api(path: str, attempts: int = 3):
    request = urllib.request.Request(
        API + path,
        headers={"user-agent": "ai-agent-stack", "accept": "application/vnd.github+json",
                 **({"authorization": f"Bearer {os.environ['GITHUB_TOKEN']}"} if os.environ.get("GITHUB_TOKEN") else {})},
    )
    for attempt in range(attempts):
        try:
            with urllib.request.urlopen(request, timeout=30) as response:
                return json.load(response)
        except urllib.error.HTTPError:
            raise
        except Exception:
            # Search is rate-limited to 30/minute and drops connections when pushed; back off and retry.
            if attempt == attempts - 1:
                raise
            time.sleep(3 * (attempt + 1))


CACHE = ROOT / "data" / ".repo-cache.json"
_cache: dict[str, dict] = json.loads(CACHE.read_text()) if CACHE.exists() else {}


def repo_info(repo: str) -> dict | None:
    """Cached: a sweep re-reads hundreds of repositories, and most of them have not changed."""
    if repo not in _cache:
        try:
            _cache[repo] = api(f"/repos/{repo}")
        except urllib.error.HTTPError:
            _cache[repo] = {}
        except Exception:
            return None
    return _cache[repo] or None


def days_since(timestamp: str) -> int:
    return max(0, (NOW - datetime.fromisoformat(timestamp.replace("Z", "+00:00"))).days)


def pool() -> dict[str, set[str]]:
    """
    Candidates, remembering which layer's query found each one.

    Everything already on the page goes in first. Without that, a listed project falls off silently the
    week its search ranking dips: the first run of this version lost Claude Code, LangGraph, LiteLLM,
    Mastra, AutoGen, Langfuse and E2B, none of which had changed at all.
    """
    found: dict[str, set[str]] = {}
    seeds = json.loads((ROOT / "seeds.json").read_text())
    for category in seeds["categories"]:
        for entry in category["repos"]:
            repo = entry if isinstance(entry, str) else entry["repo"]
            found.setdefault(repo, set()).add("listed")
    for layer, (_, _, queries, _) in LAYERS.items():
        for query in queries:
            try:
                result = api(f"/search/repositories?q={urllib.parse.quote(query + ' pushed:>' + (NOW.replace(day=1)).strftime('%Y-%m-%d'))}&sort=stars&per_page=20")
            except Exception as error:  # noqa: BLE001 - a dropped search is a gap, not a crash
                print(f"  ! {query}: {error}", file=sys.stderr)
                continue
            time.sleep(1)  # the search endpoint allows 30 requests a minute
            for item in result.get("items", []):
                found.setdefault(item["full_name"], set()).add(layer)
    for source in SOURCE_LISTS:
        try:
            with urllib.request.urlopen(urllib.request.Request(f"https://raw.githubusercontent.com/{source}/HEAD/README.md", headers={"user-agent": "x"}), timeout=30) as response:
                text = response.read().decode("utf8", "ignore")
            for owner, name in re.findall(r"github\.com/([A-Za-z0-9_.-]+)/([A-Za-z0-9_.-]+)", text):
                found.setdefault(f"{owner}/{name.rstrip('.')}", set()).add("from-list")
        except Exception as error:  # noqa: BLE001 - a missing list is a gap, not a failure
            print(f"  ! {source}: {error}", file=sys.stderr)
    return found


def user_issues(repo: str, days: int = 30) -> int:
    """Issues opened by someone who is not a maintainer. The second door into the list."""
    try:
        issues = api(f"/repos/{repo}/issues?state=all&sort=created&direction=desc&per_page=60")
    except Exception:  # noqa: BLE001 - an unreadable tracker is not evidence either way
        return 0
    return sum(
        1 for issue in issues
        if not issue.get("pull_request")
        and issue.get("author_association") in {"NONE", "CONTRIBUTOR", "FIRST_TIME_CONTRIBUTOR", "FIRST_TIMER"}
        and days_since(issue["created_at"]) <= days
    )


def classify(info: dict, hinted: set[str]) -> str | None:
    text = f"{info['full_name']} {info.get('description') or ''} {' '.join(info.get('topics') or [])}".lower()
    scored = [(len(re.findall(LAYERS[layer][3], text)), layer) for layer in LAYERS]
    best_count, best = max(scored)
    if best_count:
        return best
    real = hinted - {"from-list"}
    return sorted(real)[0] if real else None


def main() -> None:
    minimum = int(sys.argv[sys.argv.index("--stars") + 1]) if "--stars" in sys.argv else 100
    seeds = json.loads((ROOT / "seeds.json").read_text())
    already = {
        (entry if isinstance(entry, str) else entry["repo"])
        for category in seeds["categories"]
        for entry in category["repos"]
    }

    candidates = {repo: hints for repo, hints in pool().items() if not EXCLUDE.search(repo)}
    print(f"pool: {len(candidates)} candidates, reading repositories...")
    with ThreadPoolExecutor(max_workers=8) as runner:
        list(runner.map(repo_info, candidates))
    CACHE.write_text(json.dumps(_cache))
    print(f"read {len(_cache)} repositories (cached in {CACHE.name})\n")

    proposals: dict[str, list[dict]] = {}
    for repo, hints in candidates.items():
        info = repo_info(repo)
        if not info:
            continue
        reasons = []
        # The bar is now a floor, not a filter: scoring ranks, so gatekeeping by size only hides things.
        # It keeps out what cannot be measured or is already dead, and nothing else.
        adopted = info["stargazers_count"] >= minimum
        if days_since(info["pushed_at"]) > 90:
            reasons.append("not pushed in 90d")
        if info.get("fork"):
            reasons.append("fork")
        if not (info.get("license") or {}).get("spdx_id", "").replace("NOASSERTION", ""):
            reasons.append("no open licence")
        if EXCLUDE.search(info.get("description") or ""):
            reasons.append("list or tutorial")
        if "listed" in hints:
            pass
        elif not adopted:
            # A project with fewer stars but real users beats a starred repository nobody files against.
            # dapr/dapr-agents has 743 stars and 13 user issues in 30 days; stars alone would drop it.
            if user_issues(repo) < 10:
                continue
        if reasons:
            continue
        text = f"{info['full_name']} {info.get('description') or ''} {' '.join(info.get('topics') or [])}"
        if not DOMAIN.search(text):
            continue
        layer = classify(info, hints)
        if not layer:
            continue
        proposals.setdefault(layer, []).append({
            "repo": repo,
            "stars": info["stargazers_count"],
            "pushed_days": days_since(info["pushed_at"]),
            "description": (info.get("description") or "")[:80],
            "seeded": repo in already,
            "hints": sorted(hints),
        })

    out = {"generated_at": NOW.isoformat(), "bar": {"min_stars_or_user_issues": [minimum, 10], "pushed_within_days": 90, "open_licence": True}, "layers": {}}
    for layer, (title, blurb, _, _) in LAYERS.items():
        rows = sorted(proposals.get(layer, []), key=lambda row: -row["stars"])
        out["layers"][layer] = {"title": title, "blurb": blurb, "candidates": rows}
        new = [row for row in rows if not row["seeded"]]
        print(f"## {title}  ({len(rows)} pass the bar, {len(new)} not yet seeded)")
        for row in rows[:12]:
            mark = " " if row["seeded"] else "+"
            print(f" {mark} {row['repo']:<42}{row['stars']:>9,}  {row['description'][:58]}")
        print()
    (ROOT / "data" / "candidates.json").write_text(json.dumps(out, indent=1) + "\n")
    print(f"written to data/candidates.json")


if __name__ == "__main__":
    main()

File: fastagent/lib/discover.ts
import { existsSync } from "node:fs";
import { readFile, writeFile } from "node:fs/promises";
import { resolve } from "node:path";
import { HttpError, api, daysSince, json, pooled, saveEtags, searchRepos, serial, type Repo } from "./github.ts";
import { LAYERS } from "./layers.ts";

/**
 * Where candidates come from. Four sources, because each one is blind to what the others see:
 *
 * 1. **What is already listed.** Without this a listed project falls off the week its search ranking
 *    dips; a run of this pipeline once lost Claude Code, LangGraph, LiteLLM, Mastra and Langfuse, none
 *    of which had changed at all.
 * 2. **Per-layer searches**, which find projects that describe themselves in the layer's own words.
 * 3. **Brand-new repositories**, which no ranking or list has caught up with yet. This is the only
 *    source that can see a project published this week.
 * 4. **Lists other people maintain full time.** Breadth for free; measured freshness on 2026-09-13 was
 *    17/25 and 13/25 of sampled links pushed within 30 days, against 5/24 for a stale one we dropped.
 */
export const SOURCE_LISTS = [
  "hesreallyhim/awesome-claude-code",
  "kyrolabs/awesome-agents",
  "punkpeye/awesome-mcp-servers",
  "e2b-dev/awesome-sdks-for-ai-agents",
  "awesome-opencode/awesome-opencode",
  "Shubhamsaboo/awesome-llm-apps",
  "steven2358/awesome-generative-ai",
];
/** Ecosystem-wide sweeps that belong to no single layer. */
export const ECOSYSTEM_QUERIES = [
  "topic:llmops",
  "topic:agentic-ai",
  "topic:autonomous-agents",
  "topic:ai-agent stars:>300",
  "topic:ai-agents",
  "topic:llm-apps",
  "topic:generative-ai tools",
  "topic:openai stars:>500",
];
/**
 * New repositories. One window sorted by stars is one ranking, and a 45-day window ranked by stars is
 * a list of things that are 40 days old — anything published on Tuesday is crowded out before it can
 * gather a star. Several windows each get their own ranking, and the `updated` sort catches what has
 * no stars at all yet.
 */
export const FRESH_WINDOWS: { days: number; sort: string }[] = [
  { days: 7, sort: "stars" },
  { days: 21, sort: "stars" },
  { days: 60, sort: "stars" },
  { days: 7, sort: "updated" },
];
export const FRESH_QUERIES = [
  "agent in:name,description",
  "llm in:name,description",
  "mcp in:name,description",
  "ai agent in:readme",
  "agentic in:name,description",
  "claude in:name,description",
  "codex OR cursor in:name,description",
  "ai tool in:description",
];
/**
 * Per list, not shared: one counter was consumed in order, so awesome-mcp-servers and four others
 * contributed exactly zero while the first two lists spent it all.
 *
 * This is a cost ceiling, not a quality judgement — every name costs one repository read here and one
 * model judgement later — so it is named as one and set high enough that a list's own curation, not
 * our budget, decides what gets seen.
 */
const NEW_PER_LIST = Number(process.env.STACK_LIST_BUDGET ?? 400);

/** A list, a course or a demo is not a project you put in production. */
const EXCLUDE = /awesome|tutorial|course|roadmap|handbook|cookbook|examples?$|demo|starter|template|boilerplate|papers?$|interview|study|learn/i;

export type Candidate = {
  repo: string;
  stars: number;
  pushedDays: number;
  ageDays: number;
  description: string;
  foundVia: string[];
  seeded: boolean;
};

export function repoMentions(text: string): Map<string, number> {
  const counts = new Map<string, number>();
  for (const [, owner, name] of text.matchAll(/github\.com\/([A-Za-z0-9_.-]+)\/([A-Za-z0-9_.-]+)/g)) {
    if (["orgs", "topics", "features", "settings", "sponsors"].includes(owner)) continue;
    const repo = `${owner}/${name.replace(/\.git$|\.$/, "")}`;
    counts.set(repo, (counts.get(repo) ?? 0) + 1);
  }
  return counts;
}

/**
 * The floor, which is deliberately dull. Scoring ranks; a filter only hides. It keeps out what cannot
 * be measured (no licence), what is already dead, and forks. A listed project skips it entirely: only
 * a judgement removes something from the index, never a ranking wobble.
 */
export function passesFloor(repo: Repo, listed: boolean, minStars: number, userIssues = 0): boolean {
  if (listed) return true;
  if (repo.fork || repo.archived) return false;
  if (!(repo.license?.spdx_id ?? "").replace("NOASSERTION", "")) return false;
  if ((daysSince(repo.pushed_at) ?? 999) > 90) return false;
  if (EXCLUDE.test(repo.full_name) || EXCLUDE.test(repo.description ?? "")) return false;
  return repo.stargazers_count >= minStars || userIssues >= 10;
}

async function userIssueCount(repo: string): Promise<number> {
  try {
    const issues = await api<any[]>(`/repos/${repo}/issues?state=all&sort=created&direction=desc&per_page=60`);
    return issues.filter((issue) => !issue.pull_request && (daysSince(issue.created_at) ?? 999) <= 30).length;
  } catch {
    return 0;
  }
}

/**
 * Where a project is talked about before GitHub ranking notices it. Each source is wrapped: one that
 * will not load is a gap in breadth, never a reason to abandon the sweep.
 *
 * Product Hunt was checked and left out — its feed carries launches, and launches rarely link a
 * repository — and Lobsters is included cheaply despite thin GitHub coverage.
 */
export async function mentionSources(): Promise<Map<string, string>> {
  const found = new Map<string, string>();
  const add = (repo: string, source: string) => {
    if (!found.has(repo)) found.set(repo, source);
  };
  const linked = (text: string | null | undefined) => text?.match(/github\.com\/([A-Za-z0-9_.-]+\/[A-Za-z0-9_.-]+)/)?.[1];

  const month = Math.floor(Date.now() / 1000) - 30 * 86_400;
  // Ten keywords at a lower floor: a project posted to HN at 6 points is still earlier evidence than
  // anything GitHub ranking will show, and 16 hits from four keywords was leaving the source unused.
  const hnQueries = ["agent", "llm", "mcp", "ai coding", "agentic", "coding agent", "llm tool", "ai memory", "prompt", "show hn ai"];
  for (const query of hnQueries) {
    try {
      const result = await json<{ hits: { url?: string; title: string; points: number }[] }>(
        `https://hn.algolia.com/api/v1/search?tags=story&query=${encodeURIComponent(query)}&numericFilters=points>5,created_at_i>${month}&hitsPerPage=50`,
      );
      for (const hit of result.hits) {
        const repo = linked(hit.url);
        if (repo) add(repo, "hn");
      }
    } catch {
      // Algolia is free and unauthenticated; when it is down we simply see less this run.
    }
  }

  try {
    // GitHub publishes no trending API, so this reads the page. Fragile by nature, hence wrapped.
    for (const since of ["daily", "weekly"]) {
      const response = await fetch(`https://github.com/trending?since=${since}`, {
        headers: { "user-agent": "Mozilla/5.0 (ai-agent-stack)" },
        signal: AbortSignal.timeout(20_000),
      });
      if (!response.ok) continue;
      for (const [, repo] of (await response.text()).matchAll(/href="\/([\w.-]+\/[\w.-]+)\/stargazers"/g)) add(repo, `trending-${since}`);
    }
  } catch {
    // ignore
  }

  // Lobsters was tried and removed: its AI tag produced zero repository links across a full sweep.
  return found;
}

/**
 * A one-time baseline. The daily sweep only sees what was pushed in the last month or created in the
 * last two, so a project built two years ago, last touched six weeks ago and never listed anywhere is
 * invisible to it — not rejected, never seen. This enumerates the whole domain instead.
 *
 * GitHub returns at most 1,000 results per query however many pages you ask for, so the space is cut
 * into star bands until each band fits under that ceiling. The floor's own rules are pushed into the
 * query (at least 100 stars, pushed within 90 days) because filtering in the search costs one call and
 * filtering afterwards costs one call per repository.
 */
const BACKFILL_TERMS = [
  "(ai agent OR llm OR mcp OR agentic) in:name,description",
  "(assistant OR chatbot) llm in:name,description",
  "(prompt OR skills OR context) llm in:name,description",
  "(rag OR embeddings OR vector) in:name,description",
  "(openai OR anthropic OR claude OR gemini) in:name,description",
  "(智能体 OR 大模型 OR 提示词) in:name,description",
];
/** Narrow enough that each band stays under the 1,000-result ceiling for the broadest term. */
const STAR_BANDS = [
  "100..119", "120..149", "150..189", "190..249", "250..329", "330..449", "450..699",
  "700..1199", "1200..2499", "2500..4999", "5000..9999", "10000..29999", ">=30000",
];

export async function backfill(workspace: string, minStars: number): Promise<number> {
  const pool = await collectPool(workspace);
  const pushed = new Date(Date.now() - 90 * 86_400_000).toISOString().slice(0, 10);
  let added = 0;
  for (const term of BACKFILL_TERMS) {
    for (const band of STAR_BANDS) {
      const stars = band.startsWith(">=") ? `stars:${band}` : `stars:${band}`;
      const repos = await searchRepos(`${term} ${stars} pushed:>${pushed}`, "stars", 100, 10).catch(() => []);
      for (const repo of repos) {
        if (!pool.has(repo.full_name)) added++;
        pool.set(repo.full_name, (pool.get(repo.full_name) ?? new Set()).add("backfill"));
      }
      console.log(`  ${band.padEnd(12)} ${repos.length.toString().padStart(4)} results  ${term.slice(0, 42)}`);
      await new Promise((done) => setTimeout(done, 1000)); // search allows 30 requests a minute
    }
  }
  await measurePool(workspace, pool, minStars);
  return added;
}

export async function collectPool(workspace: string): Promise<Map<string, Set<string>>> {
  const found = new Map<string, Set<string>>();
  const add = (repo: string, source: string) => found.set(repo, (found.get(repo) ?? new Set()).add(source));

  const seedsPath = resolve(workspace, "seeds.json");
  if (existsSync(seedsPath)) {
    const seeds = JSON.parse(await readFile(seedsPath, "utf8")) as { categories: { repos: (string | { repo: string })[] }[] };
    for (const category of seeds.categories) {
      for (const entry of category.repos) add(typeof entry === "string" ? entry : entry.repo, "listed");
    }
  }

  const month = new Date(Date.now() - 30 * 86_400_000).toISOString().slice(0, 10);

  await serial(LAYERS.flatMap((layer) => layer.queries.map((query) => ({ layer: layer.id, query }))), 1000, async ({ layer, query }) => {
    for (const repo of await searchRepos(`${query} pushed:>${month}`, "stars", 20, 2).catch(() => [])) add(repo.full_name, layer);
  });
  await serial(ECOSYSTEM_QUERIES, 1000, async (query) => {
    for (const repo of await searchRepos(`${query} pushed:>${month}`, "updated", 30).catch(() => [])) add(repo.full_name, "ecosystem");
  });
  for (const window of FRESH_WINDOWS) {
    const since = new Date(Date.now() - window.days * 86_400_000).toISOString().slice(0, 10);
    await serial(FRESH_QUERIES, 1000, async (query) => {
      for (const repo of await searchRepos(`${query} created:>${since}`, window.sort, 30).catch(() => [])) {
        add(repo.full_name, `new-${window.days}d-${window.sort}`);
      }
    });
  }

  for (const [repo, source] of await mentionSources()) add(repo, source);

  for (const source of SOURCE_LISTS) {
    let budget = NEW_PER_LIST;
    try {
      const response = await fetch(`https://raw.githubusercontent.com/${source}/HEAD/README.md`, { signal: AbortSignal.timeout(30_000) });
      if (!response.ok) continue;
      // Most-mentioned first: a list that names a project twice is pointing at it.
      const mentioned = [...repoMentions(await response.text()).entries()].sort((a, b) => b[1] - a[1]);
      for (const [repo] of mentioned) {
        if (found.has(repo) || budget <= 0) continue;
        add(repo, `list:${source}`);
        budget -= 1;
      }
    } catch {
      // A list that will not load is a gap in breadth, not a reason to abandon the sweep.
    }
  }
  return found;
}

export async function discover(workspace: string, minStars: number) {
  return measurePool(workspace, await collectPool(workspace), minStars);
}

async function measurePool(workspace: string, pool: Map<string, Set<string>>, minStars: number) {
  const listedSet = new Set([...pool].filter(([, sources]) => sources.has("listed")).map(([repo]) => repo));
  const names = [...pool.keys()].filter((repo) => !EXCLUDE.test(repo));

  const rows = await pooled(names, 8, async (repo): Promise<Candidate | undefined> => {
    let info: Repo;
    try {
      info = await api<Repo>(`/repos/${repo}`);
    } catch (error) {
      if (error instanceof HttpError) return undefined;
      return undefined;
    }
    const listed = listedSet.has(repo);
    // The issue count is a second door into the list, so it is only worth paying for when stars fail.
    const issues = !listed && info.stargazers_count < minStars ? await userIssueCount(repo) : 0;
    if (!passesFloor(info, listed, minStars, issues)) return undefined;
    return {
      repo: info.full_name,
      stars: info.stargazers_count,
      pushedDays: daysSince(info.pushed_at) ?? 999,
      ageDays: daysSince(info.created_at) ?? 0,
      description: (info.description ?? "").slice(0, 120),
      foundVia: [...(pool.get(repo) ?? [])],
      seeded: listed,
    };
  });

  const candidates = rows.filter((row): row is Candidate => Boolean(row)).sort((a, b) => b.stars - a.stars);
  const payload = {
    generatedAt: new Date().toISOString(),
    floor: { minStars, orUserIssues30d: 10, pushedWithinDays: 90, openLicence: true },
    pool: pool.size,
    candidates,
  };
  await writeFile(resolve(workspace, "data/candidates.json"), `${JSON.stringify(payload, null, 1)}\n`);
  // Discovery never saved its ETags, so every sweep re-read all 6,500 repositories at full price and
  // spent the hourly limit several times over. A 304 costs nothing; not storing the tag costs everything.
  await saveEtags();
  return payload;
}
